flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -540,39 +571,41 @@ private FinalApplicationStatus getYarnStatus(ApplicationStatus status) {
 	/**
 	 * Request new container if pending containers cannot satisfy pending slot requests.
 	 */
-	private void requestYarnContainerIfRequired() {
-		int requiredTaskManagers = getNumberRequiredTaskManagers();
-
-		if (requiredTaskManagers > numPendingContainerRequests) {
-			requestYarnContainer();
-		}
+	private void requestYarnContainerIfRequired(Resource containerResource) {
+		getPendingWorkerNums().entrySet().stream()
+			.filter(entry ->
+				getContainerResource(entry.getKey()).equals(containerResource) &&
+				entry.getValue() > pendingWorkerCounter.getNum(entry.getKey()))
+			.findAny()
```

Comment:
Shouldn't we do this for all instead of any?

Reply:
I think this is the same issue as `KubernetesResourceManager#requestKubernetesPodIfRequired`. I was trying to keep the exactly same behavior as before. But I think you're right, it is nice to request all required pods early.